root/static/scripts/recording/components/RecordingEditForm.js

```diff
@@ -0,0 +1,496 @@
+} from '../../edit/components/ArtistCreditEditor/types.js';
+import {
+  incompleteArtistCreditFromState,
+} from '../../edit/components/ArtistCreditEditor/utilities.js';
+import EnterEdit from '../../edit/components/EnterEdit.js';
+import EnterEditNote from '../../edit/components/EnterEditNote.js';
+import FieldErrors from '../../edit/components/FieldErrors.js';
+import FormRow from '../../edit/components/FormRow.js';
+import FormRowCheckbox from '../../edit/components/FormRowCheckbox.js';
+import FormRowNameWithGuessCase, {
+  type ActionT as NameActionT,
+  runReducer as runNameReducer,
+} from '../../edit/components/FormRowNameWithGuessCase.js';
+import {NonHydratedFormRowTextList as FormRowTextList}
+  from '../../edit/components/FormRowTextList.js';
+import FormRowTextLong from '../../edit/components/FormRowTextLong.js';
+import {
+  type StateT as GuessCaseOptionsStateT,
+  createInitialState as createGuessCaseOptionsState,
+} from '../../edit/components/GuessCaseOptions.js';
+import {
+  _ExternalLinksEditor,
+  ExternalLinksEditor,
+  prepareExternalLinksHtmlFormSubmission,
+} from '../../edit/externalLinks.js';
+import guessFeat from '../../edit/utility/guessFeat2.js';
+import {
+  applyAllPendingErrors,
+  hasSubfieldErrors,
+} from '../../edit/utility/subfieldErrors.js';
+import {
+  NonHydratedRelationshipEditorWrapper as RelationshipEditorWrapper,
+} from '../../relationship-editor/components/RelationshipEditorWrapper.js';
+
+/* eslint-disable ft-flow/sort-keys */
+type ActionT =
+  | {+type: 'guess-feat'}
+  | {+type: 'show-all-pending-errors'}
+  | {+type: 'toggle-bubble', +bubble: string}
+  | {+type: 'update-name', +action: NameActionT}
+  | {+type: 'update-artist-credit', +action: ArtistCreditActionT};
+/* eslint-enable ft-flow/sort-keys */
+
+type StateT = {
+  +artistCredit: ArtistCreditStateT,
+  +form: RecordingFormT,
+  +guessCaseOptions: GuessCaseOptionsStateT,
+  +isGuessCaseOptionsOpen: boolean,
+  +recording: RecordingT,
+  +shownBubble: string,
+};
+
+function createInitialState(
+  form: RecordingFormT,
+  $c: SanitizedCatalystContextT,
+) {
+  const recording = getSourceEntityDataForRelationshipEditor($c);
+  invariant(recording && recording.entityType === 'recording');
+
+  return {
+    artistCredit: createArtistCreditState({
+      entity: recording,
+      formName: form.name,
+      id: 'source',
+    }),
+    form,
+    guessCaseOptions: createGuessCaseOptionsState(),
+    isGuessCaseOptionsOpen: false,
+    recording,
+    shownBubble: '',
+  };
+}
+
+function reducer(state: StateT, action: ActionT): StateT {
+  const newStateCtx = mutate(state);
+
+  match (action) {
+    {type: 'update-name', const action} => {
+      const nameStateCtx = mutate({
+        field: state.form.field.name,
+        guessCaseOptions: state.guessCaseOptions,
+        isGuessCaseOptionsOpen: state.isGuessCaseOptionsOpen,
+      });
+      runNameReducer(nameStateCtx, action);
+
+      const nameState = nameStateCtx.read();
```

Comment:
Prefer `nameStateCtx.final();` if you intend to not modify the value anymore.

Reply:
Seems sensible, I changed it for all four forms that did the same in a separate commit.